packager/mux/mux_test.go

```diff
@@ -217,3 +224,79 @@ func TestParamsIncorrectValueType(t *testing.T) {
 	// Expect Params to handle invalid input gracefully.
 	assert.Equal(t, Params(req), map[string]string{})
 }
+
+func TestPrometheusMetrics(t *testing.T) {
+	promTotalRequests.Reset()
+	expectedMetrics := `
+		# HELP total_requests_by_code_and_url Total number of requests by HTTP code and URL.
+		# TYPE total_requests_by_code_and_url counter
+		`
+
+	// Make requests to mux with all handlers being NOPs returning 200.
+	// Request Healthz twice to test aggregation of identical requests.
+	nopHandler := http.Handler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
+	mux := New(nopHandler, nopHandler, nopHandler, nopHandler, nopHandler)
+	pkgt.Get(t, mux, expand(`$HOST/priv/doc?fetch=$FETCH&sign=$SIGN`))
+	pkgt.Get(t, mux, expand(`$HOST/amppkg/cert/$CERT`))
+	pkgt.Get(t, mux, expand(`$HOST/amppkg/validity`))
+	pkgt.Get(t, mux, expand(`$HOST/healthz`))
+	pkgt.Get(t, mux, expand(`$HOST/healthz`))
+	expectedMetrics += `
+		total_requests_by_code_and_url{code="200",handler="signer"} 1
+		total_requests_by_code_and_url{code="200",handler="certCache"} 1
+		total_requests_by_code_and_url{code="200",handler="validityMap"} 1
+		total_requests_by_code_and_url{code="200",handler="healthz"} 2
+		`
+
+	// Test counting requests to same handler that returned different codes.
+	// Trigger a 404 attributed to healthz by adding an unexpected suffix to path.
+	pkgt.Get(t, mux, expand(`$HOST/healthzSOME_SUFFIX`))
+	expectedMetrics += `
+		total_requests_by_code_and_url{code="404",handler="healthz"} 1
```

Comment:
Optionally, break each of these stanzas up into a separate test, where each one starts with a brand new mux and ends with a call to promtest.CollectAndCompare. This has a few benefits:
  1. It's easier to diagnose failures if they're more limited in scope.
  2. We won't have any false successes from some earlier action causing a later expectation to be met.

You could use [suite](https://godoc.org/github.com/stretchr/testify/suite) to factor out the setup/teardown.

The benefits seem pretty minor in this case, but also the change seems pretty easy, so /shrug. Your call.

Reply:
Thanks for the input. I agree with the pros of singling out each scenario.

My intention was to run an "integration" test to check that when being accumulated, different metrics don't interfere with each other.

To achieve both goals I've put all the scenarios in an array and implemented running them both in isolation and altogether.